### PR TITLE
feat: add allowFreeInput to v8 ComboBox to allow typing while the combobox is in focus

### LINKED ISSUE #19379: ComboBox autocomplete does not work
### Environment Information

- @fluentui/react 8.26.2

### Please provide a reproduction of the bug in a codepen:

I'm using the combo-box and trying get autocomplete behaviour to work. See the code sandbox for the stark difference between the MaterialUI ComboBox which works flawlessly and the fluent ui version. The textbox part of the ComboBox also looks broken without allowFreeform as you can start to type but when a selection is made backspace cannot be used and effectively locks in the selection blocking autocomplete.

https://codesandbox.io/s/combo-box-comparison-ldmlw

User Demo:
AutoComplete

#### Actual behavior:

The ComboBox only opens when the user starts typing, and does not filter the options as they type.

#### Expected behavior:

The ComboBox should open as soon as the control is focused and the options should be filtered as the user starts to enter text.

### Priorities and help requested:

Are you willing to submit a PR to fix? I do not know enough to answer.

Requested priority: Normal? It does seem to indicate autocomplete does not work.

Products/sites affected: Product In development.

### PR DESCRIPTION
**Pull request by smhigley**
Resolves #19379

## Current Behavior

The current v8 combobox will never allow the user to enter text that does not match an existing option unless `allowFreeform=true`. This leaves out use cases where the combobox should not allow freeform text strings to be submitted, but allowing free user search while open/focused is still desired.

This also addresses the many quirks around autocomplete without allowFreeform.

## New Behavior

If `allowFreeInput` is set to `true`, ComboBox allows freeform typing and backspace within the input. When the field is blurred, the value resets to the last selected value, if the string does not match an option. This matches the current behavior of the field if you type with `allowFreeform=false` without opening the listbox for the first time -- the only time free typing was allowed previously.

Example 

https://fluentuipr.z22.web.core.windows.net/pull/25602/react/storybook/index.html?path=/story/components-combobox--free-input

### CODE REVIEW

File: packages/react/src/components/ComboBox/ComboBox.tsx [hunk 1]
```diff
@@ -554,12 +555,7 @@ class ComboBoxInternal extends React.Component<IComboBoxInternalProps, IComboBox
       return null;
     }
 
-    const visibleValue = normalizeToString(this._currentVisibleValue);
-    if (comboBox.value !== visibleValue) {
-      return visibleValue;
-    }
-
-    return comboBox.value;
+    return normalizeToString(this._currentVisibleValue);
```
Comment (smhigley) [inline packages/react/src/components/ComboBox/ComboBox.tsx:558]: This is unrelated to the other changes in the PR, but I came across it while investigating the issue.

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/d5d03853c4aabcf7cedd715caddd987ba44c1a60'>d5d03853c4aabcf7cedd715caddd987ba44c1a60</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/react/src/components/ComboBox/ComboBox.types.ts [hunk 2]
```diff
@@ -138,6 +138,13 @@ export interface IComboBoxProps
    */
   allowFreeform?: boolean;
 
+  /**
+   * When true, the Combobox will allow the user to type freely while the Combobox is focused.
+   * On Blur, the value will be set to the matching option, or the previous selection.
```
Comment (khmakoto) [inline packages/react/src/components/ComboBox/ComboBox.types.ts:143]:
```suggestion
   * On Blur, the value will be set to the matching option, or the previous selection if there is no match.
```

File: packages/react/src/components/ComboBox/ComboBox.tsx [hunk 3]
```diff
@@ -388,7 +389,7 @@ class ComboBoxInternal extends React.Component<IComboBoxInternalProps, IComboBox
             prevProps.hoisted.selectedIndices &&
             selectedIndices &&
             prevProps.hoisted.selectedIndices[0] !== selectedIndices[0]) ||
-            !allowFreeform ||
+            (!allowFreeform && !allowFreeSearch) ||
```
Comment (khmakoto) [inline packages/react/src/components/ComboBox/ComboBox.tsx:392]: You should either update or remove the comment about this if (lines 376-382).

File: packages/react/src/components/ComboBox/ComboBox.tsx [hunk 4]
```diff
@@ -2183,7 +2180,7 @@ class ComboBoxInternal extends React.Component<IComboBoxInternalProps, IComboBox
         // If we are not allowing freeform and are not autoComplete
```
Comment (khmakoto) [inline packages/react/src/components/ComboBox/ComboBox.tsx:2180]:
```suggestion
        // If we are not allowing freeform nor freeinput, and autoComplete is off
```

File: packages/react-examples/src/react/ComboBox/ComboBox.FreeSearch.Example.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import * as React from 'react';
+import { ComboBox, IComboBoxOption, IComboBoxStyles, SelectableOptionMenuItemType } from '@fluentui/react';
```
Comment (khmakoto) [inline packages/react-examples/src/react/ComboBox/ComboBox.FreeSearch.Example.tsx:2]:
```suggestion
import { ComboBox, SelectableOptionMenuItemType } from '@fluentui/react';
import type { IComboBoxOption, IComboBoxStyles } from '@fluentui/react';
```

File: packages/react-examples/src/react/ComboBox/ComboBox.FreeSearch.Example.tsx [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        allowFreeSearch
+        autoComplete="off"
+      />
+      <ComboBox
+        defaultSelectedKey="C"
+        label="ComboBox with allowFreeSearch without autocomplete"
+        options={options}
+        styles={comboBoxStyles}
+        allowFreeSearch
+        autoComplete="on"
```
Comment (khmakoto) [inline packages/react-examples/src/react/ComboBox/ComboBox.FreeSearch.Example.tsx:38]: I think these examples have their label's swapped

```suggestion
        label="ComboBox with allowFreeSearch and autocomplete"
        options={options}
        styles={comboBoxStyles}
        allowFreeSearch
        autoComplete="on"
      />
      <ComboBox
        defaultSelectedKey="C"
        label="ComboBox with allowFreeSearch without autocomplete"
        options={options}
        styles={comboBoxStyles}
        allowFreeSearch
        autoComplete="off"
```

File: packages/react/src/components/ComboBox/ComboBox.test.tsx [hunk 7]
```diff
@@ -589,6 +609,58 @@ describe('ComboBox', () => {
...
+    const { getByRole } = render(<ComboBox options={DEFAULT_OPTIONS2} allowFreeSearch onChange={onChange} />);
+    const combobox = getByRole('combobox');
+    userEvent.type(combobox, 'foo');
+    expect(onChange).not.toHaveBeenCalled();
+
+    userEvent.type(combobox, '{enter}');
+    expect(onChange).toHaveBeenCalledWith(expect.anything(), DEFAULT_OPTIONS2[1], 1, 'Foo');
+  });
+
+  it('onChange is not called when a non-matching is typed and submitted with allowFreeSearch', () => {
```
Comment (khmakoto) [inline packages/react/src/components/ComboBox/ComboBox.test.tsx:644]:
```suggestion
  it('onChange is not called when a non-matching text is typed and submitted with allowFreeSearch', () => {
```

File: packages/react-examples/src/react/ComboBox/ComboBox.FreeSearch.Example.tsx (hunk 6 above)
Comment (smhigley) [reply packages/react-examples/src/react/ComboBox/ComboBox.FreeSearch.Example.tsx:38]: haha good catch, I completely missed that 🙈

**Review by khmakoto — APPROVED** [review]

File: packages/react-examples/src/react/ComboBox/ComboBox.FreeInput.Example.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  { key: 'D', text: 'Option D' },
+  { key: 'divider', text: '-', itemType: SelectableOptionMenuItemType.Divider },
+  { key: 'Header2', text: 'Second heading', itemType: SelectableOptionMenuItemType.Header },
+  { key: 'E', text: 'Option E' },
+  { key: 'F', text: 'Option F', disabled: true },
+  { key: 'G', text: 'Option G' },
+  { key: 'H', text: 'Option H' },
+  { key: 'I', text: 'Option I' },
+  { key: 'J', text: 'Option J' },
+];
```
Comment (JustSlone) [inline packages/react-examples/src/react/ComboBox/ComboBox.FreeInput.Example.tsx:19]: When all the options start with "Option" it makes trying out the auto-complete-freeinput behavior tricky. I'd recommend something that has more variation between options.

**Comment by fabricteam** [comment]
## [🕵 fluentuiv8 Open the Visual Regressions report to inspect the 7 screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-tool-fluent/index.html?organization=uifabric&project=fabricpublic&prId=25602&commitId=0c43505b553e7426cb4632d39bd400ceed515863&env=prod)

✅ There was 0 screenshots added, 4 screenshots removed, 1032 screenshots unchanged, 0 screenshots with different dimensions and 3 screenshots with visible difference.


  <details><summary><b>unknown</b> 7 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|TagPicker.Root.chromium.png|151| Changed|
|TagPicker.Selected - RTL.chromium.png|151| Changed|
|TagPicker.Selected.chromium.png|151| Changed|
|Pivot - Overflow.Root.Narrow - Last tab selected.chromium.png|0| Removed|
|Pivot - Overflow.Root.Narrow - Overflow menu.chromium.png|0| Removed|
|Pivot - Overflow.Tabs - RTL.Narrow - Last tab selected.chromium.png|0| Removed|
|Pivot - Overflow.Tabs - RTL.Narrow - Overflow menu.chromium.png|0| Removed|

</details>


<div id="vrtCommentfluentuiv8"/>

**Review by micahgodbolt — APPROVED** [review]

**Pull request merged**